sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+text of the failing expression, the values of the standard macros `__FILE__` and
+`__LINE__`, and the value of the standard variable `__func__`. If the failing
+assert comes from an `nd_range` `parallel_for` it will also include the global
+ID and the local ID of the failing work item.
+
+Implementation of this function is supplied by Native Device Compiler for
+safe approach or by DPCPP Compiler for fallback one.
+
+In order to distinguish which implementation to use, DPCPP Runtime checks for
+`cl_intel_devicelib_cassert` extension. If the extension isn't available, then
+fallback implementation is used.
+
+
+## Safe approach
+
+This is the preferred approach and implementations should use it when possible.
+It guarantees assertion failure notification delivery to the host regardless of
+kernel behavior which hit the assertion.
+
+The Native Device Compiler is responsible for providing implementation of
+`__devicelib_assert_fail` which completely hides details of communication
+between the device code and the Low-Level Runtime from the SYCL device compiler
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+      // It's not defined if GPU thread stops execution
+      // Other GPU threads are left untouched
+   2. Specialized version of `__devicelib_assert_fail` is called
+   3. Device immediately signals to host (Low-Level Runtime)
+ - Host side:
+   1. The assert failure gets detected by Low-Level Runtime
+   2. Low-Level Runtime prints assert failure message to `stderr`
+   3. Low-Level Runtime calls `abort()`
+
+
+## Fallback approach
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a fallback
+approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a mutable program scope variable is introduced. This
+variable stores a flag which says if an assert failure was encountered. Fallback
+implementation of `__devicelib_assert_fail` atomically raises the flag so that
+DPCPP Runtime is able to detect assert failure after kernel finishes.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+   2. Fallback version of `__devicelib_assert_fail` is called
+   3. Assert information is stored into program-scope variable
+   4. Kernel continues running
+ - Host side:
+   1. A copy 'kernel' is enqueued as the one depending on user's kernel to get
+      the value of assert failure flag.
+   2. A host-task is enqueued to check value of assert failure flag.
+   3. The host task calls abort whenever assert failure flag is set.
+
```

Comment:
For device side it is clear that with NDEBUG there is no overhead. For host side it is not quite. From reading below I can see that the 3 items above happen only if the device code has asserts enabled, so:
```suggestion
DPCPP Runtime will automatically check if assertions are enabled in the kernel being run, and won't enqueue the auxiliary kernels if assertions are not enabled. So there is no host-side runtime overhead when assertion are not enabled.
```